### PR TITLE
Fix zk session race condition.

### PR DESCRIPTION
**Pull request by huizhilu**
This PR draft is for initial review purpose. I am still working on completing unit tests, adding more comments, and running tests.

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/manager/zk/ZkSessionAwareStateListener.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import org.I0Itec.zkclient.IZkStateListener;
+
+
+public interface ZkSessionAwareStateListener extends IZkStateListener {
```
Comment (huizhilu) [inline helix-core/src/main/java/org/apache/helix/manager/zk/ZkSessionAwareStateListener.java]: Will add comments for this public interface.

Comment (lei-xia) [reply helix-core/src/main/java/org/apache/helix/manager/zk/ZkSessionAwareStateListener.java]: Why do we need this?

File: helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java [hunk 2]
```diff
@@ -521,6 +543,11 @@ public String create(final String path, Object data, final CreateMode mode)
    */
   public String create(final String path, Object datat, final List<ACL> acl, final CreateMode mode)
       throws IllegalArgumentException, ZkException {
+    return create(path, datat, acl, mode, null);
+  }
+
+  private String create(final String path, Object datat, final List<ACL> acl, final CreateMode mode,
```
Comment (lei-xia) [inline helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java]: need JavaDoc to explain what is usage of sessionId here.

File: helix-core/src/main/java/org/apache/helix/manager/zk/ZkSessionAwareStateListener.java (hunk 1 above)
Comment (lei-xia) [reply helix-core/src/main/java/org/apache/helix/manager/zk/ZkSessionAwareStateListener.java]: If we have to introduce this new method, why not add one method into existing IZkStateListener (with default implementation just calling the old one in the interface definition)?

**Comment by lei-xia** [comment]
Can we use  more formal PR for reviewing?  This is not a good way to track all people's comments.

File: helix-core/src/main/java/org/apache/helix/manager/zk/ZkSessionAwareStateListener.java (hunk 1 above)
Comment (huizhilu) [reply helix-core/src/main/java/org/apache/helix/manager/zk/ZkSessionAwareStateListener.java]: We need a new api: handleNewSession(sessionId). 

handleNewSession() is defined in IZkStateListener, which is an interface defined in I0Itec.zkclient. We can not just change handleNewSession() with handleNewSession(sessionId) in I0Itec.zkclient. Considering backward compatibility, we need to extend interface IZkStateListener to add a new api: handleNewSession(sessionId). The benefit is subscriptions of  IZkStateListener don't have to change.

So with this new extended interface, we can both support backward compatibility and have no changes to stateListener subscriptions.

File: helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java (hunk 2 above)
Comment (huizhilu) [reply helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java]: Agreed. I've added detailed java doc for this api in my local branch.

File: helix-core/src/main/java/org/apache/helix/manager/zk/ZkSessionAwareStateListener.java (hunk 1 above)
Comment (lei-xia) [reply helix-core/src/main/java/org/apache/helix/manager/zk/ZkSessionAwareStateListener.java]: I think we are getting rid of I0Itec when we are moving all ZK code to the new zkclient module. If that is the case, should we wait for that to make the interface clear (and less confusing)?

**Comment by huizhilu** [comment]
> Can we use more formal PR for reviewing? This is not a good way to track all people's comments.

@lei-xia My initial idea was to have an internal review before make it public on Apache Helix repo. To track all people's comments, I guess you meant notifications? You can still subscribe notifications on the right bar like below:
image

Once the internal review is completed, I will make corresponding changes and submit a PR to Apache Helix. Does it sound good? If you prefer the draft to be on Apache Helix now, I could definitely submit one.

File: helix-core/src/main/java/org/apache/helix/manager/zk/ZkSessionAwareStateListener.java (hunk 1 above)
Comment (huizhilu) [reply helix-core/src/main/java/org/apache/helix/manager/zk/ZkSessionAwareStateListener.java]: Regarding this, we may discuss further. I think that would be a good idea to wait for that, if we can wait for that in reasonable time.

**Comment by lei-xia** [comment]
No, I am not meaning the notification.  I am saying we need a place to keep track the history of discussions and agreements on resolutions.  You do not have to get a perfect PR before submitting to Apache Helix reviewing, that is the purpose of PR reviewing.

**Comment by huizhilu** [comment]
> No, I am not meaning the notification. I am saying we need a place to keep track the history of discussions and agreements on resolutions. You do not have to get a perfect PR before submitting to Apache Helix reviewing, that is the purpose of PR reviewing.

@lei-xia Thanks for the clarification. I will submit a draft PR to Apache Helix.

**Pull request closed without merging**